akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSubscriber.scala

```diff
@@ -0,0 +1,120 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl.io
+
+import java.util.concurrent.BlockingQueue
+
+import akka.actor._
+import akka.stream.actor.{ ActorSubscriberMessage, WatermarkRequestStrategy }
+import akka.stream.impl.io.InputStreamSubscriber.Failed
+import akka.util.ByteString
+
+/** INTERNAL API */
+object InputStreamSubscriber {
+  def props(sharedBuffer: BlockingQueue[ByteString]): Props = {
+    Props(new InputStreamSubscriber(sharedBuffer)).withDeploy(Deploy.local)
+  }
+
+  trait InputStreamSubscriberInMessage
+
+  case object Request extends InputStreamSubscriberInMessage
+  case object ReadNotification extends InputStreamSubscriberInMessage
+  case object Close extends InputStreamSubscriberInMessage
+
+  trait InputStreamSubscriberOutMessage
+
+  case class AcknowledgeMessage(terminated: Boolean) extends InputStreamSubscriberOutMessage
+  case object Finished extends InputStreamSubscriberOutMessage
+  case class Failed(cause: Throwable) extends InputStreamSubscriberOutMessage
+
+}
+
+/** INTERNAL API */
+private[akka] class InputStreamSubscriber(sharedBuffer: BlockingQueue[ByteString])
+  extends akka.stream.actor.ActorSubscriber with ActorLogging {
+
+  val bufSize = sharedBuffer.remainingCapacity()
+
+  var bytesRequestedDownstream: Option[ActorRef] = None
+  var bytesSentDownstream: Long = 0
+  var receivedError: Option[Throwable] = None
+
+  var onCompleteReceived = false
+  var sentCompleteNotification = false
+
+  override protected val requestStrategy = WatermarkRequestStrategy(highWatermark = bufSize)
+
+  def receive = {
+    case ActorSubscriberMessage.OnNext(bytes: ByteString) ⇒
+      if (bytes.nonEmpty) {
+        // ignore empty ByteStrings
+        val added = sharedBuffer.offer(bytes)
+        assert(added, "Bytes must always be added to buffer as adapter consuming them")
```

Comment:
please don't use assert, they produce AssertionErrors which is not handled by NonFatal, and therefore bubbles right through the whole actor hierarchy with the final effect of killing the ActorSystem. Please use require() instead.